### PR TITLE
Backend/feat/#3357/edit pickup

### PR DESCRIPTION
**Pull request by JanhaviSoni**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Update.hs [hunk 1]
```diff
@@ -59,3 +74,41 @@ handler req = do
   logTagInfo "Payment completed : " ("bookingId " <> req.bookingId.getId <> ", rideId " <> req.rideId.getId)
 
 -- TODO store to DB
+
+handler EditLocationReq {..} = do
+  ride <- QRide.findById rideId >>= fromMaybeM (RideNotFound rideId.getId)
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Update.hs:79]: run in Replica

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Update.hs [hunk 2]
```diff
@@ -59,3 +74,41 @@ handler req = do
...
+  person <- QPerson.findById ride.driverId >>= fromMaybeM (PersonNotFound ride.driverId.getId)
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Update.hs:80]: run in Replica

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Update.hs [hunk 3]
```diff
@@ -59,3 +74,41 @@ handler req = do
...
+          DL.LocationAddress
+            { street = location.address.street,
+              door = location.address.door,
+              city = location.address.city,
+              state = location.address.state,
+              country = location.address.country,
+              building = location.address.building,
+              areaCode = location.address.area_code,
+              area = location.address.locality,
+              fullAddress = Nothing
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Update.hs:112]: earlier also, are we sending fullAddress as Nothing ?
Can't we concat everything and construct fullAddress ?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/LocationMapping.hs [hunk 4]
```diff
@@ -25,6 +25,8 @@ import Kernel.Types.Id
 import qualified Sequelize as Se
 import qualified Storage.Beam.LocationMapping as BeamLM
 
+-- import qualified Storage.Beam.Location as DL
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/LocationMapping.hs:28]: remove commented code

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs [hunk 5]
```diff
@@ -137,3 +164,109 @@ getRideStatus rideId personId = withLogTag ("personId-" <> personId.getId) do
...
+  ) =>
+  Id SRide.Ride ->
+  (Id SPerson.Person, Id DM.Merchant) ->
+  EditLocationReq ->
+  m APISuccess
+editLocation rideId (_, merchantId) req = do
+  when (isNothing req.origin && isNothing req.destination) do
+    throwError PickupOrDropLocationNotFound
+
+  ride <- QRide.findById rideId >>= fromMaybeM (RideNotFound rideId.getId)
```
Comment (RitikaHanish) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs:184]: runInReplica

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs [hunk 6]
```diff
@@ -137,3 +164,109 @@ getRideStatus rideId personId = withLogTag ("personId-" <> personId.getId) do
...
+
+  let attemptsLeft = fromMaybe 1 ride.allowedEditLocationAttempts
+  when (attemptsLeft == 0) do
+    throwError EditLocationAttemptsExhausted
+
+  let bookingId = ride.bookingId
+  booking <- QRB.findById bookingId >>= fromMaybeM (BookingNotFound bookingId.getId)
```
Comment (RitikaHanish) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs:191]: runInReplica

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs [hunk 7]
```diff
@@ -137,3 +164,109 @@ getRideStatus rideId personId = withLogTag ("personId-" <> personId.getId) do
...
+  (Id SPerson.Person, Id DM.Merchant) ->
+  EditLocationReq ->
+  m APISuccess
+editLocation rideId (_, merchantId) req = do
+  when (isNothing req.origin && isNothing req.destination) do
+    throwError PickupOrDropLocationNotFound
+
+  ride <- QRide.findById rideId >>= fromMaybeM (RideNotFound rideId.getId)
+
+  let attemptsLeft = fromMaybe 1 ride.allowedEditLocationAttempts
```
Comment (RitikaHanish) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs:186]: In this way if ride.allowedEditLocationAttempts is Nothing, then customer can edit location infinite no if times.
Correct your logic.

**Review by RitikaHanish — CHANGES_REQUESTED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs (hunk 7 above)
Comment (JanhaviSoni) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs:186]: Please check line no 235 , attempts are getting updated so 1 will get updated to 0

File: Backend/lib/beckn-spec/src/Beckn/Types/Core/Taxi/Update/UpdateEvent/UpdateEventType.hs [hunk 8]
```diff
@@ -19,12 +19,14 @@ import Kernel.Prelude
...
 data UpdateEventType
   = PAYMENT_COMPLETED
+  | EDIT_LOCATION
   deriving (Show, Eq, Ord, Read, Generic, ToSchema)
 
 -- Generic instances for type with single value will not work
 instance FromJSON UpdateEventType where
   parseJSON (String "PAYMENT_COMPLETED") = pure PAYMENT_COMPLETED
-  parseJSON (String _) = parseFail "Expected \"PAYMENT_COMPLETED\""
+  parseJSON (String "EDIT_LOCATION") = pure EDIT_LOCATION
```
Comment (roman-bodavskiy) [inline Backend/lib/beckn-spec/src/Beckn/Types/Core/Taxi/Update/UpdateEvent/UpdateEventType.hs:28]: you can use `deriving (FromJSON, ToJSON)` for `UpdateEventType`
it works wrong only for enumerations with single item like `data SingleValue = SingleValue`

File: Backend/lib/beckn-spec/src/Beckn/Types/Core/Taxi/Update/UpdateEvent/EditLocationEvent.hs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+          & required
+            L..~ [ "id",
+                   "./komn/update_target",
+                   "fulfillment"
+                 ]
+
+data FulfillmentInfo = FulfillmentInfo
+  { id :: Text,
+    origin :: Maybe Location,
+    destination :: Maybe Location
```
Comment (roman-bodavskiy) [inline Backend/lib/beckn-spec/src/Beckn/Types/Core/Taxi/Update/UpdateEvent/EditLocationEvent.hs:92]: does not fit to beckn spec
https://app.swaggerhub.com/apis/ONDC/ONDC-Mobility-API-Specification/1.0.0-draft#/Beckn%20Provider%20Platform%20(BPP)/post_update
can we use the same fulfillment as in `Beckn.Types.Core.Taxi.Common.FulfillmentInfo` (maybe remove not required fields)?

**Review by roman-bodavskiy — CHANGES_REQUESTED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs [hunk 10]
```diff
@@ -307,6 +307,7 @@ onUpdate ValidatedRideAssignedReq {..} = do
             rideEndTime = Nothing,
             rideRating = Nothing,
             safetyCheckStatus = Nothing,
+            allowedEditLocationAttempts = Nothing,
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs:310]: It should not be Nothing, it should take some value from config table.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Update.hs [hunk 11]
```diff
@@ -59,3 +78,51 @@ handler req = do
...
+
+mkFullAddress :: Common.Address -> Maybe Text
+mkFullAddress Common.Address {..} = do
+  let strictFields = catMaybes $ filter (not . isEmpty) [door, building, street, locality, city, state, area_code, country]
+  if null strictFields
+    then Nothing
+    else Just $ T.intercalate ", " strictFields
+
+isEmpty :: Maybe Text -> Bool
+isEmpty = maybe True (T.null . T.replace " " "")
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Update.hs:128]: It should be `T.null . T.strip`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Update.hs [hunk 12]
```diff
@@ -59,3 +78,51 @@ handler req = do
...
 
 -- TODO store to DB
+
+handler EditLocationReq {..} = do
+  ride <- runInReplica $ QRide.findById rideId >>= fromMaybeM (RideNotFound rideId.getId)
+  person <- runInReplica $ QPerson.findById ride.driverId >>= fromMaybeM (PersonNotFound ride.driverId.getId)
+  whenJust origin $ \pickup -> do
+    startLocation <- buildLocation pickup
+    QL.create startLocation
+    pickupMapForBooking <- SLM.buildPickUpLocationMapping startLocation.id bookingId.getId DLM.BOOKING (Just person.merchantId) (Just person.merchantOperatingCityId)
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Update.hs:88]: Why are we changing the booking? In this case on ride table will be changed, right?

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs (hunk 7 above)
Comment (hkmangla) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs:187]: If value Nothing then default should come from same config as added in previous comment.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs [hunk 13]
```diff
@@ -138,3 +165,109 @@ getRideStatus rideId personId = withLogTag ("personId-" <> personId.getId) do
...
+
+  let bookingId = ride.bookingId
+  booking <- B.runInReplica $ QRB.findById bookingId >>= fromMaybeM (BookingNotFound bookingId.getId)
+  merchant <- CQMerchant.findById merchantId >>= fromMaybeM (MerchantNotFound merchantId.getId)
+
+  whenJust req.origin $ \pickup -> do
+    when (ride.status /= SRide.NEW) do
+      throwError (InvalidRequest "Customer is not allowed to change pickup as the ride is not NEW")
+    lastMapping <- QLM.findLastMapping ride.id.getId 0 >>= fromMaybeM (InternalError "Lastest mapping not found")
+    initialLocationForRide <- QL.findById lastMapping.locationId >>= fromMaybeM (InternalError "Location not found")
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Ride.hs:199]: Proper error handling.

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Queries/Merchant.hs [hunk 14]
```diff
@@ -94,7 +94,9 @@ instance FromTType' BeamM.Merchant Merchant where
             aadhaarKeyExpiryTime = aadhaarKeyExpiryTime,
             mediaFileSizeUpperLimit = mediaFileSizeUpperLimit,
             mediaFileUrlPattern = mediaFileUrlPattern,
-            trackingShortUrlPattern = trackingShortUrlPattern
+            trackingShortUrlPattern = trackingShortUrlPattern,
+            editPickupDistanceThreshold = fromMaybe 100 editPickupDistanceThreshold,
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Storage/Queries/Merchant.hs:98]: Why to have it as Maybe value, it could be non Maybe value.

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Queries/Merchant.hs (hunk 14 above)
Comment (JanhaviSoni) [reply Backend/app/rider-platform/rider-app/Main/src/Storage/Queries/Merchant.hs:98]: Yes I kept it non maybe only but due to Drainer we need to keep new columns as maybe

**Review by hkmangla — DISMISSED** [review]

**Pull request merged**